### PR TITLE
Enabled toolbar focus in source editing mode

### LINKED ISSUE #10368: It's impossible to focus the toolbar with the keyboard in the source editing mode
## 📝 Provide a description of the improvement

1.  Use the keyboard to enter the source editing mode.
2.  Focus on the editable and type something.
3.  Try to move back to the toolbar with the keyboard.

Related - #9906.

---

If you'd like to see this improvement implemented, add a 👍 reaction to this post.

### PR DESCRIPTION
**Pull request by mateuszzagorski**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/contributing/git-commit-message-convention.html))

Fix (source-editing): It is possible now to focus the toolbar with the keyboard in the source editing mode. Closes #10368.

---

### Additional information

_For example – encountered issues, assumptions you had to make, other affected tickets, etc._

### CODE REVIEW

**Review by Dumluregn — CHANGES_REQUESTED** [review]
My findings:

1) The tests coverage report shows, that the code you added in the editing part of the plugin isn't tested. You can check it locally running `yarn test -f source-editing --coverage` and inspecting the report created in the `coverage` directory. It suggests something is wrong with tests.
2) If you comment the code you added, the tests are still passing. Another indicator that tests don't actually test the code they are intended to.
3) Inspecting the tests code:
- You are creating a new toolbar separated from the editor used in the test and don't attach it anyhow. That means that after `const toolbar = editor.ui.view.toolbar;`, `toolbar` is `undefined` and the keystrokes handlers are not registered.
- So why the tests pass? Let's take a look at the `enableToolbarKeyboardFocus()` function used there. Its docs say:
```
Enables focus/blur toolbar navigation using `Alt+F10` and `Esc` keystrokes.
```
So it does the same job we try to do, and it seems it does it well since the tests pass.
- The conclusion - we are using the solution we should in tests, but not in the plugin itself. We should use the dedicated helper and then remember to use the correct toolbar in the tests.


Other than that, I've noticed that the `Esc` key works only after the first switch to the source editing mode - when you switch again, it stops working. But maybe that issue will be solved by changing the implementation.

**Review by Dumluregn — APPROVED** [review]
LGTM, requesting another review from the core team. A couple of notes to whomever checks this PR later:
1) We create a separate `focusTracker` here - I'm not 100% sure if that's the right approach or we should reuse the one from the editor. 
2) AFAICS the `focusTracker.destroy()` method does not clear the list of tracked elements - on contrary to its docs: https://github.com/ckeditor/ckeditor5/blob/c41a8e6c0d920668b2b724fe7c918e964386b396/packages/ckeditor5-utils/src/focustracker.js#L109-L116. In did in the first version: https://github.com/ckeditor/ckeditor5/commit/594d6c3c80ffa58f8f3d91d0482a592e84ed9c33, but later that line was removed due to some redundancy - but I don't see it: https://github.com/ckeditor/ckeditor5/commit/950bf6c05d5bba4dd92efa06c308eb22b15f15ff. I think it should be restored - asking for the confirmation.
3) I believe the commit that I've recalled in 2) should be enough to close https://github.com/ckeditor/ckeditor5/issues/4971 - also waiting for the approval.

**Comment by wimleers** [comment]
Tracked downstream at https://www.drupal.org/project/drupal/issues/3283800#comment-14561193 👍

**Review by oleq — CHANGES_REQUESTED** [review]
Aside from the concerns expressed by @Dumluregn, I must admit that I don't like this whole approach from the architectural point of view.

By hardcoding Alt+F10 and Esc logic in `SourceEditing` we're straying further from `enableToolbarKeyboardFocus()` that was an attempt to keep things in one place. The following code works because `SourceEditing's` `KeystrokeHandler` listens to the `editor.ui.view.toolbar.element`

https://github.com/ckeditor/ckeditor5/blob/ef2565f184bed48af61b7df89d79e8fa6814ca4b/packages/ckeditor5-source-editing/src/sourceediting.js#L295-L301

And it works magically probably because it won the race in the order of execution upon the `Esc` key that supposed to be encapsulated in `enableToolbarKeyboardFocus()` (two `KeystrokeHandlers`, two listeners for the same key)

https://github.com/ckeditor/ckeditor5/blob/30286f77b39526fce2856b03b9be0ba4cc91d1c7/packages/ckeditor5-ui/src/toolbar/enabletoolbarkeyboardfocus.js#L53-L63

We should centralize the logic, not scatter it across features. Since `enableToolbarKeyboardFocus()` is not enough to handle all this (it'a 5yo helper developed when there was a single editor, no source editing etc.), what we need is a piece of code that:
* will know about the `editor.ui.view.toolbar` (for instance, `BalloonEditorUIView` does not have one, but it has something else), so probably it should be provided by the implementation of the editor, not `SourceEditing`
    * will know that in some editors, focusing toolbar is not enough, some must be shown first
* will allow registering DOM elements (editing DOM root(s), source editing's textarea, what else?)
* will offer an API to focus the toolbar (since it knows about it) that individual features' `KeystrokeHandlers` could hook up to `Alt+F10` or, even better, an API that will hide `Alt+F10` from the features altogether,
* will be able to revert focus back upon `Esc` to the place where it originally came from

Also, since we want to have Alt+F10 a focal point of navigation across different toolbars (https://github.com/ckeditor/ckeditor5/issues/5146, https://github.com/ckeditor/ckeditor5/issues/5473, etc.), the design of the new system should this requirement into account.

**Comment by mlewand** [comment]
Closing the PR in favor of #12064.

**Pull request closed without merging**
